In [10]:
import csv
import pandas as pd

In [11]:
file = "stanford_crime_clean.csv"
df = pd.read_csv(file)

In [12]:
df.columns = [col.lower().replace(" ", "_") for col in df.columns]

In [14]:
df.head()

unnamed:_0                        nature   case_#       reported      date  \
0           0                 Vehicle Theft  19-0364   3/1/19 17:10  03/01/19   
1           1            Burglary - Vehicle  19-0255  2/11/19 18:22  02/11/19   
2           2            Burglary - Vehicle  19-0365   3/2/19 12:31  03/02/19   
3           3  Inf Corp Injury:spouse/Cohab  19-0367    3/2/19 9:44  03/02/19   
4           4   Outside Warrant Arrest MISD  19-0001   1/1/19 11:44  01/01/19   

    time                               occurred  \
0  17:10  02/28/2019  19:00 -\n03/01/2019 17:10   
1  18:22              02/11/2019  15:30 - 15:35   
2  12:31              03/02/2019  15:00 - 00:15   
3  09:44                                 3/2/19   
4  11:44              01/01/2019  10:40 - 10:57   

                                            location            disposition  \
0   Mirrielees Highrise, 730 Escondido Rd,\nStanford         Cleared/Closed   
1  Golf Driving Range Facility, 51 Campus Dr,\nSt...         Cleared/Closed   
2   Ps 1 - Roth Way Garage, 345 Campus Dr,\nStanford         Cleared/Closed   
3  Munger Graduate Res -, 554 Salvatierra\nWlk, S...  Clred By Adult Arrest   
4   Ps 1 - Roth Way Garage, 345 Campus Dr,\nStanford          Cited (Adult)   

  on_campus?  area           x          y  
0        Yes  10.0 -122.159135  37.423030  
1         No   5.0 -122.184350  37.425473  
2         No   7.0 -122.172452  37.432882  
3        Yes   NaN -122.167434  37.422788  
4         No   7.0 -122.172452  37.432882

In [15]:
df.tail()

unnamed:_0                                             nature   case_#  \
4114        4114  Grand\ntheft:money/labor/property - All\nother...  23-0421   
4115        4115                                  Petty Theft: Bike  23-0423   
4116        4116                                  Petty Theft: Bike  23-0424   
4117        4117                                  Petty Theft: Bike  23-0425   
4118        4118                                  Petty Theft: Bike  23-0427   

           reported      date   time                               occurred  \
4114  3/25/23 12:50  03/25/23  12:50  03/24/2023  22:00 -\n03/25/2023 12:30   
4115  3/26/23 12:43  03/26/23  12:43  03/24/2023  19:30 -\n03/26/2023 12:30   
4116  3/26/23 13:54  03/26/23  13:54  03/25/2023  18:00 -\n05/26/2023 13:30   
4117  3/26/23 15:11  03/26/23  15:11  03/25/2023  21:00 -\n03/26/2023 15:00   
4118  3/26/23 16:58  03/26/23  16:58  03/25/2023  17:00 -\n03/26/2023 11:30   

                                               location     disposition  \
4114  Lagunita - Dining, 326 Santa Teresa St,\nStanford  Cleared/Closed   
4115               Pi Beta Phi, 517 Cowell Ln, Stanford  Cleared/Closed   
4116               Pi Beta Phi, 517 Cowell Ln, Stanford  Cleared/Closed   
4117  Rains Apts - Bldg 225, 225 Ayrshire Farm\nLn, ...  Cleared/Closed   
4118        Kappa Alpha Theta, 585 Cowell Ln,\nStanford  Cleared/Closed   

     on_campus?  area           x          y  
4114         No   5.0 -122.174073  37.422324  
4115         No  10.0 -122.161898  37.421457  
4116         No  10.0 -122.161898  37.421457  
4117         No  10.0 -122.159515  37.421823  
4118         No  10.0 -122.162017  37.421243

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4119 entries, 0 to 4118
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   unnamed:_0   4119 non-null   int64  
 1   nature       4119 non-null   object 
 2   case_#       4119 non-null   object 
 3   reported     4119 non-null   object 
 4   date         4119 non-null   object 
 5   time         4115 non-null   object 
 6   occurred     4119 non-null   object 
 7   location     4119 non-null   object 
 8   disposition  4096 non-null   object 
 9   on_campus?   4119 non-null   object 
 10  area         2840 non-null   float64
 11  x            4119 non-null   float64
 12  y            4119 non-null   float64
dtypes: float64(3), int64(1), object(9)
memory usage: 418.5+ KB


In [17]:
df = df.astype({"nature":"string"})

In [18]:
df = df.astype({"location":"string"})

In [19]:
df = df.astype({"disposition":"string"})

In [20]:
df = df.astype({"on_campus?":"string"})

In [21]:
check_col = df[pd.isnull(df["nature"])]
pd.set_option('display.max_columns', None)
check_col

Empty DataFrame
Columns: [unnamed:_0, nature, case_#, reported, date, time, occurred, location, disposition, on_campus?, area, x, y]
Index: []

In [22]:
check_col = df[pd.isnull(df["disposition"])]
pd.set_option('display.max_columns', None)
check_col

unnamed:_0                                             nature   case_#  \
3003        3003                         Warrant Arrest Misdemeanor  22-0363   
4014        4014  False personation of
another:spec circ - False...  23-0444   
4034        4034  Annoying telephone
call:obscene/threatening -
...  23-0484   
4037        4037                    Petty theft - From building (M)  23-0488   
4038        4038  Student Safety 
Battery:spouse/ex
spouse/date/...  23-0489   
4039        4039                                  Petty Theft: Bike  23-0490   
4041        4041  Student Safety 
Assault to commit rape -
Simpl...  23-0494   
4042        4042                                  Petty Theft: Bike  23-0496   
4043        4043                    Petty theft - From building (M)  23-0497   
4044        4044                       Vandalism less than $400 (M)  23-0498   
4045        4045                                  Petty Theft: Bike  23-0499   
4046        4046                               Grand theft:bike (F)  23-0507   
4047        4047                      Appropriate lost property (M)  23-0509   
4048        4048                                  Petty Theft: Bike  23-0511   
4049        4049                                  Petty Theft: Bike  23-0512   
4051        4051                                  Petty Theft: Bike  23-0518   
4053        4053                       Unauth entry occu dwling (M)  23-0523   
4054        4054                                  Petty Theft: Bike  23-0530   
4057        4057                                  Petty Theft: Bike  23-0536   
4059        4059       Burglary:2nd degree - From
motor vehicle (F)  23-0545   
4060        4060  Obtain money/etc by false
pretenses [over $400...  23-0546   
4061        4061  Obtain money/etc by false
pretenses [over $400...  23-0548   
4063        4063                                  Petty Theft: Bike  23-0552   

           reported      date   time                                occurred  \
3003    4/1/22 9:47  04/01/22  09:47                             4/1/22 9:41   
4014   3/31/23 8:26  03/31/23  08:26               03/28/2023  14:44 - 14:44   
4034   4/6/23 12:10  04/06/23  12:10               04/05/2023  13:14 - 13:15   
4037   4/6/23 19:11  04/06/23  19:11   04/06/2023  13:30 -\n04/06/2023 19:00   
4038   4/6/23 20:15  04/06/23  20:15                            4/6/23 20:15   
4039   4/6/23 16:59  04/06/23  16:59   03/24/2023  17:00 -\n04/04/2023 10:00   
4041   4/7/23 10:30  04/07/23  10:30  03/25/2023 - 04/02/2023,\nUnknown Time   
4042   4/7/23 15:15  04/07/23  15:15   04/06/2023  18:30 -\n04/07/2023 13:00   
4043   4/7/23 17:15  04/07/23  17:15   04/07/2023  13:30 -\n04/07/2023 15:30   
4044    4/8/23 7:17  04/08/23  07:17   04/01/2023  14:00 -\n04/08/2023 07:00   
4045   4/8/23 16:32  04/08/23  16:32   04/02/2023  11:00 -\n06/06/2023 11:00   
4046   4/9/23 17:31  04/09/23  17:31   03/24/2023  12:00 -\n04/09/2023 16:30   
4047   4/9/23 21:27  04/09/23  21:27   04/08/2023  12:00 -\n04/08/2023 14:00   
4048  4/10/23 14:13  04/10/23  14:13   04/09/2023  20:30 -\n04/09/2023 22:15   
4049  4/10/23 15:39  04/10/23  15:39   04/10/2023  01:50 -\n04/10/2023 12:00   
4051  4/11/23 15:53  04/11/23  15:53   04/07/2023  16:30 -\n04/10/2023 16:45   
4053   4/13/23 7:32  04/13/23  07:32               04/06/2023  00:30 - 00:45   
4054  4/13/23 20:08  04/13/23  20:08               04/11/2023  16:30 - 16:40   
4057  4/14/23 15:30  04/14/23  15:30   04/11/2023  17:30 -\n04/13/2023 10:20   
4059  4/17/23 12:21  04/17/23  12:21               04/17/2023  10:49 - 12:18   
4060  4/17/23 12:00  04/17/23  12:00   04/12/2023  08:32 -\n04/15/2023 02:00   
4061  4/17/23 17:30  04/17/23  17:30   04/09/2023  12:00 -\n04/14/2023 12:00   
4063  4/18/23 11:38  04/18/23  11:38                                 4/18/23   

                                               location disposition  \
3003                       Panama Ml/Duena St, Stanford        <NA>   
4014               Evg

It seems like there's a decent amount of entries where the disposition is unknown (is the case closed, active? etc.) This could be cleared up with some additonal reporting and asking the lieutenant (if we have time). 

In [23]:
check_col = df[pd.isnull(df["on_campus?"])]
pd.set_option('display.max_columns', None)
check_col

Empty DataFrame
Columns: [unnamed:_0, nature, case_#, reported, date, time, occurred, location, disposition, on_campus?, area, x, y]
Index: []

In [27]:
check_col = df[pd.isnull(df["area"])]
pd.set_option('display.max_columns', None)
check_col

unnamed:_0                                             nature   case_#  \
3              3                       Inf Corp Injury:spouse/Cohab  19-0367   
5              5                                  Petty Theft: bike  19-0261   
7              7                                 Burglary - Vehicle  19-0368   
13            13                                  Drive W/O License  19-0462   
17            17                                 Burglary - Vehicle  19-0371   
...          ...                                                ...      ...   
4097        4097  Petty theft - From motor vehicle
(M) 
Vandalis...  23-0393   
4098        4098                                  Petty Theft: Bike  23-0394   
4100        4100      Petty Theft: Bike 
Warrant Arrest Misdemeanor  23-0397   
4104        4104       Trespass:occupy property
without consent (M)  23-0402   
4112        4112     Attempted Petty theft - From
motor vehicle (M)  23-0417   

           reported      date   time                               occurred  \
3       3/2/19 9:44  03/02/19  09:44                                 3/2/19   
5     2/13/19 13:25  02/13/19  13:25  02/12/2019  13:00 -\n12/12/2019 16:00   
7      3/2/19 13:26  03/02/19  13:26              03/02/2019  12:00 - 13:00   
13     3/17/19 2:53  03/17/19  02:53              03/17/2019  00:30 - 00:35   
17     3/3/19 15:38  03/03/19  15:38              03/03/2019  12:00 - 13:47   
...             ...       ...    ...                                    ...   
4097  3/20/23 19:02  03/20/23  19:02  03/19/2023  20:30 -\n03/20/2023 18:40   
4098  3/20/23 17:48  03/20/23  17:48  03/10/2023  12:00 -\n03/13/2023 17:00   
4100   3/21/23 8:30  03/21/23  08:30                           3/21/23 8:30   
4104  3/22/23 19:17  03/22/23  19:17                          3/22/23 19:18   
4112   3/25/23 2:49  03/25/23  02:49                          3/24/23 22:30   

                                               location  \
3     Munger Graduate Res -, 554 Salvatierra
Wlk, St...   
5       Huang Engineering Ctr, 475 Via Ortega,
Stanford   
7              (Hundred Block.), .300 Palm Dr, Stanford   
13                            373 Oxford Ave, Palo Alto   
17            (Hundred Block.), .400 Roth Way, Stanford   
...                                                 ...   
4097            Parking Lot 40, 675 Lomita Dr, Stanford   
4098                   Lantana, 750 Campus Dr, Stanford   
4100      (Hundred Block.), .700 Escondido Rd,
Stanford   
4104  450 Jane Stanford Way Bldg 380, 450 Jane
Stanf...   
4112     Lot 51 - 1035 Campus, 1035 Campus Dr,
Stanford   

                disposition on_campus?  area           x          y  
3     Clred By Adult Arrest        Yes   NaN -122.167434  37.422788  
5            Cleared/Closed         No   NaN -122.174376  37.427910  
7            Cleared/Closed         No   NaN -122.168201  37.433549  
13            Cited (Adult)         No   NaN -122.146783  37.428435  
17           Cleared/Closed         No   NaN -122.167348  37.431068  
...                     ...        ...   ...         ...        ...  
4097         Cleared/Closed         No   NaN -122.174073  37.422324  
4098         Cleared/Closed         No   NaN -122.160843  37.425402  
4100      Cleared by Arrest         No   NaN -122.160290  37.424179  
4104      Cleared by Arrest         No   NaN -122.170060  37.427660  
4112         Cleared/Closed         No   NaN -122.170259  37.420252  

[1279 rows x 13 columns]

In [24]:
df.groupby("case_#").min()

unnamed:_0                                             nature  \
case_#                                                                   
19-0001           4                        Outside Warrant Arrest MISD   
19-0002          10                                  Battery On Person   
19-0003          16                         Warrant Arrest Misdemeanor   
19-0004          23                          Drive W/License Suspended   
19-0007          77                                 Burglary - Vehicle   
...             ...                                                ...   
23-0545        4059       Burglary:2nd degree - From
motor vehicle (F)   
23-0546        4060  Obtain money/etc by false
pretenses [over $400...   
23-0548        4061  Obtain money/etc by false
pretenses [over $400...   
23-0550        4062  Disorderly conduct:lodge without
owners consen...   
23-0552        4063                                  Petty Theft: Bike   

              reported      date   time  \
case_#                                    
19-0001   1/1/19 11:44  01/01/19  11:44   
19-0002   1/1/19 17:55  01/01/19  17:55   
19-0003   1/2/19 17:32  01/02/19  17:32   
19-0004   1/2/19 20:53  01/02/19  20:53   
19-0007   1/3/19 12:35  01/03/19  12:35   
...                ...       ...    ...   
23-0545  4/17/23 12:21  04/17/23  12:21   
23-0546  4/17/23 12:00  04/17/23  12:00   
23-0548  4/17/23 17:30  04/17/23  17:30   
23-0550   4/18/23 2:15  04/18/23  02:15   
23-0552  4/18/23 11:38  04/18/23  11:38   

                                      occurred  \
case_#                                           
19-0001              01/01/2019  10:40 - 10:57   
19-0002              01/01/2019  16:10 - 16:24   
19-0003              01/02/2019  16:50 - 17:00   
19-0004              01/02/2019  20:08 - 20:08   
19-0007              01/03/2019  11:05 - 11:10   
...                                        ...   
23-0545              04/17/2023  10:49 - 12:18   
23-0546  04/12/2023  08:32 -\n04/15/2023 02:00   
23-0548  04/09/2023  12:00 -\n04/14/2023 12:00   
23-0550                           4/18/23 2:15   
23-0552                                4/18/23   

                                                  location        disposition  \
case_#                                                                          
19-0001    Ps 1 - Roth Way Garage, 345 Campus Dr,
Stanford      Cited (Adult)   
19-0002  Black Comm Svcs Ctr, 418 Santa Teresa St,
Stan...     Cleared/Closed   
19-0003  (Hundred Block.), .800 Junipero Serra Blvd,
St...      Cited (Adult)   
19-0004    Ps 1 - Roth Way Garage, 345 Campus Dr,
Stanford      Cited (Adult)   
19-0007          (Hundred Block.), .300 Roth Way, Stanford     Cleared/Closed   
...                                                    ...                ...   
23-0545      Lot 95 - Track House, 295 Galvez St,
Stanford               <NA>   
23-0546  Residence - 715 Salvatierra St, 715
Salvatierr...               <NA>   
23-0548  Elliott Program Center, 589 Governors
Ave, Sta...               <NA>   
23-0550  450 Jane Stanford Way Bldg 200, 450 Jane
Stanf...  Cleared by Arrest   
23-0552       620 Mayfield Ave, 620 Mayfield Ave,
Stanford               <NA>   

        on_campus?  area           x          y  
case_#                                           
19-0001         No   7.0 -122.172452  37.432882  
19-0002         No   2.0 -122.171116  37.425141  
19-0003         No  12.0 -122.161017  37.410099  
19-0004         No   7.0 -122.172452  37.432882  
19-0007         No   NaN -122.168838  37.431370  
...            ...   ...         ...        ...  
23-0545         No   9.0 -122.163632  37.431553  
23-0546         No  11.0 -122.162676  37.418913  
23-0548         No   5.0 -122.178058  37.424256  
23-0550         No   NaN -122.170060  37.427660  
23-0552         No   NaN -122.166542  37.419369  

[4119 rows x 12 columns]

In [28]:
check_col = df[pd.isnull(df["location"])]
pd.set_option('display.max_columns', None)
check_col

Empty DataFrame
Columns: [unnamed:_0, nature, case_#, reported, date, time, occurred, location, disposition, on_campus?, area, x, y]
Index: []

In [25]:
df.groupby("nature").min()

/var/folders/n0/8zv0xmjx2tlfg6cn8s0jgktm0000gn/T/ipykernel_90198/2669167530.py:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.min is deprecated. In a future version, a TypeError will be raised. Before calling .min, select only columns which should be valid for the function.
  df.groupby("nature").min()


unnamed:_0   case_#  \
nature                                                                   
Abandoned Vehicle                                         260  19-0041   
Acquire or possess access card of                         790  19-0817   
another                                                  3985  22-1666   
Pet...                                                   2913  22-0650   
Acquire/possess personal id info                          638  19-0814   
w/intent defr...                                          970  19-2010   
Animal Bite, Identifying Info                            3504  22-1220   
Required                                                 3355  22-1350   
Annoy/Etc Child -18                                      3894  23-0151   
Warrant Arrest Misdemeanor                               2724  21-0693   
Possess Controlled...                                     ...      ...   
Wear/etc id/etc:impers po - False                                        
pretenses/sw...                                                          
...                                                                      
Wilfl dschrg bb gn gr neg -                                              
Aggravated (M)                                                           
Wilfl dschrg bb gn gr neg -                                              
Weapon law violati...                                                    
Willful Cruelty To Child                                                 

                                                         reported      date  \
nature                                                                        
Abandoned Vehicle                                   1/10/19 17:19  01/10/19   
Acquire or possess access card of                    5/9/19 15:59  05/09/19   
another                                            11/22/22 15:30  11/22/22   
Pet...                                              5/21/22 18:32  05/21/22   
Acquire/possess personal id info                      5/9/19 8:05  05/09/19   
w/intent defr...                                    10/28/19 1:11  10/28/19   
Animal Bite, Identifying Info                       9/19/22 11:03  09/19/22   
Required                                            10/6/22 20:05  10/06/22   
Annoy/Etc Child -18                                  2/5/23 21:13  02/05/23   
Warrant Arrest Misdemeanor                          8/30/21 15:30  08/30/21   
Possess Controlled...                                         ...       ...   
Wear/etc id/etc:impers po - False                                             
pretenses/sw...                                                               
...                                                                           
Wilfl dschrg bb gn gr neg -                                                   
Aggravated (M)                                                                
Wilfl dschrg bb gn gr neg -                                                   
Weapon law violati...                                                         
Willful Cruelty To Child                                                      

                                                                                 occurred  \
nature                                                                                      
Abandoned Vehicle                                               01/10/2019  14:45 - 14:45   
Acquire or possess access card of                               05/09/2019  08:45 - 14:00   
another                                             09/01/2022  12:00 -\n09/15/2022 12:00   
Pet...                                                                      5/20/22 11:00   
Acquire/possess personal id info                   04/01/2018 - 05/09/2019,\nUnknown Time   
w/intent defr...                                                10/28/2019  00:20 - 00:50   
Animal Bite, Identifying Info                                               9/19/22 10:36   

Looks like there are some abbreviations used by the person who entered this data that we will have to clarify and edit in the dataframe to make it clear when users see it on the map.

I wonder if we should clean up the names so they're in the same categories (i.e. display all petty thefts together instead of differentiating between petty theft-bike and petty theft-store?) I am unsure about this because I don't want to misrepresent the data. 

In [26]:
df.groupby("nature").count().max()

unnamed:_0     663
case_#         663
reported       663
date           663
time           663
occurred       663
location       663
disposition    663
on_campus?     663
area           518
x              663
y              663
dtype: int64

There's 663 unique types of crimes entered in this dataframe. There's quite a few entries where there was nothing entered for the area number of where on campus it occured.

In [29]:
df["unnamed:_0"].count()

4119